error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Foo<_, _>`
  --> $DIR/conflict-unpin.rs:3:15
   |
3  | #[pin_project(UnsafeUnpin)] //~ ERROR E0119
   |               ^^^^^^^^^^^ conflicting implementation for `Foo<_, _>`
...
10 | impl<T, U> Unpin for Foo<T, U> where T: Unpin {}
   | --------------------------------------------- first implementation here
   |
   = note: upstream crates may add a new impl of trait `pin_project::UnsafeUnpin` for type `pin_project::__private::Wrapper<'_, Foo<_, _>>` in future versions

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Bar<_, _>`
  --> $DIR/conflict-unpin.rs:12:15
   |
12 | #[pin_project(UnsafeUnpin)] //~ ERROR E0119
   |               ^^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
19 | impl<T, U> Unpin for Bar<T, U> {}
   | ------------------------------ first implementation here
   |
   = note: upstream crates may add a new impl of trait `pin_project::UnsafeUnpin` for type `pin_project::__private::Wrapper<'_, Bar<_, _>>` in future versions

error[E0119]: conflicting implementations of trait `std::marker::Unpin` for type `Baz<_, _>`
  --> $DIR/conflict-unpin.rs:21:15
   |
21 | #[pin_project(UnsafeUnpin)] //~ ERROR E0119
   |               ^^^^^^^^^^^ conflicting implementation for `Baz<_, _>`
...
28 | impl<T: Unpin, U: Unpin> Unpin for Baz<T, U> {}
   | -------------------------------------------- first implementation here
   |
   = note: upstream crates may add a new impl of trait `pin_project::UnsafeUnpin` for type `pin_project::__private::Wrapper<'_, Baz<_, _>>` in future versions
