error[E0119]: conflicting implementations of trait `_::FooMustNotImplDrop` for type `Foo<_, _>`
 --> $DIR/conflict-drop.rs:5:1
  |
5 | #[pin_project] //~ ERROR E0119
  | ^^^^^^^^^^^^^^
  | |
  | first implementation here
  | conflicting implementation for `Foo<_, _>`
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `std::ops::Drop` for type `Bar<_, _>`
  --> $DIR/conflict-drop.rs:16:15
   |
16 | #[pin_project(PinnedDrop)] //~ ERROR E0119
   |               ^^^^^^^^^^ conflicting implementation for `Bar<_, _>`
...
28 | impl<T, U> Drop for Bar<T, U> {
   | ----------------------------- first implementation here
