error: #[repr(packed)] attribute should be applied to a struct or union
  --> $DIR/packed-enum.rs:13:8
   |
13 | #[repr(packed)] //~ ERROR E0517
   |        ^^^^^^

error: #[repr(packed)] attribute should be applied to a struct or union
  --> $DIR/packed-enum.rs:18:8
   |
18 | #[repr(packed)] //~ ERROR E0517
   |        ^^^^^^

error[E0517]: attribute should be applied to a struct or union
  --> $DIR/packed-enum.rs:7:8
   |
7  |   #[repr(packed)] //~ ERROR E0517
   |          ^^^^^^
8  | / enum E1 {
9  | |     V(()),
10 | | }
   | |_- not a struct or union

error[E0517]: attribute should be applied to a struct or union
  --> $DIR/packed-enum.rs:13:8
   |
13 |   #[repr(packed)] //~ ERROR E0517
   |          ^^^^^^
14 | / enum E2 {
15 | |     V(()),
16 | | }
   | |_- not a struct or union

error[E0517]: attribute should be applied to a struct or union
  --> $DIR/packed-enum.rs:18:8
   |
18 |   #[repr(packed)] //~ ERROR E0517
   |          ^^^^^^
19 |   #[pin_project]
20 | / enum E3 {
21 | |     V(()),
22 | | }
   | |_- not a struct or union
