error: expected item after attributes
 --> $DIR/packed_sneaky-3.rs:4:1
  |
4 | hidden_repr_macro! {} //~ ERROR expected item after attributes
  | ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected item after attributes
  --> $DIR/packed_sneaky-3.rs:13:9
   |
13 |         #[repr(packed)] //~ ERROR expected item after attributes
   |         ^^^^^^^^^^^^^^^
...
17 | hidden_repr_macro2! {}
   | ---------------------- in this macro invocation
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected item after attributes
  --> $DIR/packed_sneaky-3.rs:24:10
   |
24 | #[derive(HiddenRepr)] //~ ERROR expected item after attributes
   |          ^^^^^^^^^^
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: proc-macro derive produced unparseable tokens
  --> $DIR/packed_sneaky-3.rs:24:10
   |
24 | #[derive(HiddenRepr)] //~ ERROR expected item after attributes
   |          ^^^^^^^^^^
