error[E0308]: mismatched types
  --> $DIR/remove-attr-from-field.rs:28:38
   |
28 |     let _: Pin<&mut PhantomPinned> = x.f; //~ ERROR E0308
   |            -----------------------   ^^^ expected struct `Pin`, found `&mut PhantomPinned`
   |            |
   |            expected due to this
   |
   = note:         expected struct `Pin<&mut PhantomPinned>`
           found mutable reference `&mut PhantomPinned`

error[E0308]: mismatched types
  --> $DIR/remove-attr-from-field.rs:32:38
   |
32 |     let _: Pin<&mut PhantomPinned> = x.f; //~ ERROR E0308
   |            -----------------------   ^^^ expected struct `Pin`, found `&mut PhantomPinned`
   |            |
   |            expected due to this
   |
   = note:         expected struct `Pin<&mut PhantomPinned>`
           found mutable reference `&mut PhantomPinned`
